### PR TITLE
Better i18n support for emotion elements

### PR DESCRIPTION
**Pull request by pascalheidmann-bedarf**
### 1. Why is this change necessary?
The use of direct naming emotion elements is a mess. Several elements don't have a correct translation and adding own elements require to use either the name directly or hack snippets into the extjs model.
Adding another problem with the names: even the author of the docs for creating custom emotion elements don't know how they are handled as the [example](https://developers.shopware.com/developers-guide/custom-shopping-world-elements/#registering-a-new-element) will always recreate the same element.

### 2. What does this change do, exactly?
- Move snippets management to the data layer -> snippets will be fetched inside the controller action instead of template. Fetching logic is backwards compatible so plugins still can use the old code.
- renaming default shopping world elements to a more useful internal name (the already used key for snippets)
- making the `name` field of emotion elements unique
- Adding fallback solution for `$componentName` in the function `Shopware\Components\Emotion\ComponentInstaller::createOrUpdate()` so you can leave out the `name` parameter and avoid having inconsistent data and therefore problems with new elements

### 3. Describe each step to reproduce the issue or behaviour.
Use the code [from the custom shopping world example](https://developers.shopware.com/developers-guide/custom-shopping-world-elements/#registering-a-new-element) several times to have the element multiple times.

### 4. Please link to the relevant issues (if any).
https://issues.shopware.com/issues/SW-24441

### 5. Which documentation changes (if any) need to be made because of this PR?
- https://developers.shopware.com/developers-guide/custom-shopping-world-elements/#registering-a-new-element should be modified (even without this PR, see issue/this PR)
- snippet namespace should be mentioned: `backend/emotion/view/detail`
- (internal) shopping world element name has to be unique (technically per plugin but should be unique in general)

### 6. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: _sql/migrations/1635-rename-default-emotion-elements.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * trademark license. Therefore any rights, title and interest in
+ * our trademarks remain entirely with us.
+ */
+
+class Migrations_Migration1635 extends Shopware\Components\Migrations\AbstractMigration
+{
+    public function up($modus)
+    {
+        $updates = [
+            ['Artikel', 'article'],
```
Comment (shyim) [inline _sql/migrations/1635-rename-default-emotion-elements.php]: article :cry:

File: _sql/migrations/1635-rename-default-emotion-elements.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            ['Kategorie-Teaser', 'category_teaser'],
+            ['Blog-Artikel', 'blog_article'],
+            ['Banner', 'banner'],
+            ['Banner-Slider', 'banner_slider'],
+            ['Youtube-Video', 'youtube'],
+            ['Hersteller-Slider', 'manufacturer'],
```
Comment (shyim) [inline _sql/migrations/1635-rename-default-emotion-elements.php]: manufacturer_slider

File: _sql/migrations/1635-rename-default-emotion-elements.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            ['Artikel-Slider', 'article_slider'],
```
Comment (shyim) [inline _sql/migrations/1635-rename-default-emotion-elements.php]: product_slider

File: engine/Shopware/Controllers/Backend/Emotion.php [hunk 4]
```diff
@@ -446,12 +447,20 @@ public function libraryAction()
...
 
         $components = $builder->getQuery()->getArrayResult();
+
+        $snippets = $this->get('snippets')->getNamespace('backend/emotion/view/detail');
+        foreach ($components as &$component) {
+            $name = str_replace(' ', '_', strtolower($component['name']));
+
+            $component['fieldLabel'] = $snippets->get($name, $component['name']);
+        }
+
```
Comment (shyim) [inline engine/Shopware/Controllers/Backend/Emotion.php:472]: unset $component

Comment (pascalheidmann-bedarf) [reply engine/Shopware/Controllers/Backend/Emotion.php:472]: As I'm using a reference instead of a copy wouldn't `unset($component);` remove that element?

Comment (shyim) [reply engine/Shopware/Controllers/Backend/Emotion.php:472]: https://github.com/kalessil/phpinspectionsea/blob/master/docs/performance.md#foreach-variables-reference-usage-correctness :)
Maybe install their Plugin. Its awesome good :blush:

Comment (pascalheidmann-bedarf) [reply engine/Shopware/Controllers/Backend/Emotion.php:472]: Ah now I get what you mean. I added it even if the variable is scoped within the function and therefore shouldn't be problematic (as long as nobody is messing with the variables later)

File: engine/Shopware/Controllers/Backend/Emotion.php [hunk 5]
```diff
@@ -446,12 +447,21 @@ public function libraryAction()
...
 
         $components = $builder->getQuery()->getArrayResult();
+
+        $snippets = $this->get('snippets')->getNamespace('backend/emotion/view/detail');
+        foreach ($components as &$component) {
+            $name = str_replace(' ', '_', strtolower($component['name']));
+
+            $component['fieldLabel'] = $snippets->get($name, $component['name']);
+        }
+        unset($component); // to make shyim happy and prevent potential issues (scoped anyway)
```
Comment (tinect) [inline engine/Shopware/Controllers/Backend/Emotion.php:471]: 😆 best comment

**Comment by shyim** [comment]
Hey,
thanks for your Pull Request :blush: 
It looks good, but the element names in the blocks are now gone. Can you fix that?

In 5.6:
old

In your Branch:
new

**Comment by pascalheidmann-bedarf** [comment]
@shyim I guess the missing labels where the result of the migration collision when in the meantime a wild Migration1635 appeared ;)
Sadly I can't yet test it myself as my local setup is broken right now :(

**Comment by PaddyS** [comment]
Hey @pascalheidmann-bedarf,

can you please ping us once you've fixed your local setup and tested your new changes?

**Comment by pascalheidmann-bedarf** [comment]
@PaddyS @shyim found the bug that prevented the correct translation from popping up in the shopping world designer. Turned out Shopware is pushing the elements name once again in the `detailAction`

**Comment by shyim** [comment]
Thanks for your Contribution! :blue_heart: 
We have merged it into 5.7, cause the renaming could be a breaking change

**Pull request closed without merging**